src/persist/mod.rs

```diff
@@ -0,0 +1,207 @@
+                            Child::Node(node) => {
+                                Self::persist_inner(
+                                    backend,
+                                    &*node.compound()?,
+                                )?;
+                            }
+                            Child::EndOfNode => break,
+                            _ => (),
+                        }
+                    }
+                }
+                PutResult::AlreadyPresent => (),
+            }
+        }
+
+        Ok(PersistedId(id))
+    }
+}
+
+lazy_static! {
+    static ref BACKENDS: Arc<RwLock<HashMap<u64, WrappedBackend>>> =
+        Arc::new(RwLock::new(HashMap::new()));
+}
+
+/// A backend constructor
+pub struct BackendCtor<B> {
+    ctor: fn() -> B,
+    id: u64,
+}
+
+impl<B> BackendCtor<B> {
+    /// Create a new constructor from a function/closure
+    pub fn new(ctor: fn() -> B) -> Self {
+        let mut hasher = DefaultHasher::new();
+        Hash::hash(&ctor, &mut hasher);
+        let id = hasher.finish();
+        BackendCtor { ctor, id }
+    }
+}
+
+/// Id of a persisted GenericTree
+pub struct PersistedId(Id);
+
+impl PersistedId {
+    /// Restore a GenericTree from a persistance backend.
+    pub fn restore(&self) -> Result<GenericTree, PersistError> {
+        Persistance::get(&self.0)
+    }
+}
+
+/// The singleton interface to the persistance layer
+pub struct Persistance;
+
+impl Persistance {
+    /// Persist the given Compound to a backend
+    pub fn persist<C, A, B>(
+        ctor: &BackendCtor<B>,
+        c: &C,
+    ) -> Result<PersistedId, PersistError>
+    where
+        C: Compound<A>,
+        C::Leaf: Canon,
+        A: Annotation<C::Leaf>,
+        B: 'static + Backend,
+    {
+        let mut backends = BACKENDS.write();
+        let entry = backends.entry(ctor.id);
+
+        match entry {
+            Entry::Occupied(mut occupied) => occupied.get_mut().persist(c),
+            Entry::Vacant(vacant) => {
+                let backend = (ctor.ctor)();
+                vacant.insert(WrappedBackend::new(backend)).persist(c)
+            }
+        }
+    }
+
+    /// Get a generic tree from the backend.
+    pub fn get(id: &Id) -> Result<GenericTree, PersistError> {
+        // First try reifying from local store/inlined data
+        if let Ok(tree) = id.reify() {
+            return Ok(tree);
+        }
+
+        let backends = BACKENDS.read();
+
+        for (_, backend) in backends.iter() {
+            if let Ok(tree) = backend.get(id) {
+                return Ok(tree);
+            }
+        }
+        Err(CanonError::NotFound.into())
+    }
+
+    /// Returns a constructor for a temporary test backend
+    pub fn test_backend_ctor() -> BackendCtor<TestBackend> {
+        BackendCtor::new(|| {
+            let dir = tempfile::tempdir().unwrap();
+            let b = DiskBackend::new(dir.path()).unwrap();
+            TestBackend(b, dir)
+        })
+    }
```

Comment:
Test method shouldn't be part of the API, we should have it as helper function in the integration test.